### PR TITLE
Retry storage operation on 409 conflict

### PR DESCRIPTION
**Pull request by sonjaer**
# TL;DR
Fix for https://github.com/flyteorg/flyte/issues/1997

## Type
 - [x] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: pkg/common/data_store.go [hunk 1]
```diff
@@ -23,8 +25,33 @@ func OffloadLiteralMap(ctx context.Context, storageClient *storage.DataStore, li
...
+	})
+
+	if err != nil {
 		return "", errors.NewFlyteAdminErrorf(codes.Internal, "Failed to write protobuf for [%+v] with err: %v", nestedKeys, err)
 	}
+
 	return uri, nil
 }
+
+func Retry(attempts int, delay time.Duration, f func() error) error {
```
Comment (sonjaer) [inline pkg/common/data_store.go:41]: Ideally we could change and use https://github.com/flyteorg/flyteadmin/blob/master/pkg/async/shared.go#L13 to only retry on certain errors that can be passed as arguments

File: pkg/common/data_store.go [hunk 2]
```diff
@@ -23,8 +25,33 @@ func OffloadLiteralMap(ctx context.Context, storageClient *storage.DataStore, li
 	if err != nil {
 		return "", errors.NewFlyteAdminErrorf(codes.Internal, "Failed to construct data reference for [%+v] with err: %v", nestedKeys, err)
 	}
-	if err := storageClient.WriteProtobuf(ctx, uri, storage.Options{}, literalMap); err != nil {
+
+	err = Retry(5, async.RetryDelay, func() error {
```
Comment (sonjaer) [inline pkg/common/data_store.go:29]: Extract to config?

File: pkg/common/data_store.go [hunk 3]
```diff
@@ -23,8 +25,33 @@ func OffloadLiteralMap(ctx context.Context, storageClient *storage.DataStore, li
...
 }
+
+func Retry(attempts int, delay time.Duration, f func() error) error {
+	var err error
+	for attempt := 0; attempt <= attempts; attempt++ {
+		err = f()
+		if err == nil {
+			return nil
+		}
+		if e, ok := err.(*googleapi.Error); ok && e.Code == 409 {
```
Comment (sonjaer) [inline pkg/common/data_store.go:48]: A bit hard to reproduce this as my local version does not connect to GCP and even if I am not sure I could get this error to reappear.  I used this https://github.com/graymeta/stow/blob/973a61f346d598a566affb53c4698764a67df164/google/location.go#L38-L42 to understand how to handle it

File: pkg/common/data_store.go [hunk 4]
```diff
@@ -23,8 +23,22 @@ func OffloadLiteralMap(ctx context.Context, storageClient *storage.DataStore, li
 	if err != nil {
 		return "", errors.NewFlyteAdminErrorf(codes.Internal, "Failed to construct data reference for [%+v] with err: %v", nestedKeys, err)
 	}
-	if err := storageClient.WriteProtobuf(ctx, uri, storage.Options{}, literalMap); err != nil {
+
+	err = async.RetryOnSpecificErrors(5, async.RetryDelay, func() error {
```
Comment (sonjaer) [inline pkg/common/data_store.go:27]: Extract to config?

File: pkg/common/data_store.go [hunk 5]
```diff
@@ -23,8 +23,22 @@ func OffloadLiteralMap(ctx context.Context, storageClient *storage.DataStore, li
...
+
+	if err != nil {
 		return "", errors.NewFlyteAdminErrorf(codes.Internal, "Failed to write protobuf for [%+v] with err: %v", nestedKeys, err)
 	}
+
 	return uri, nil
 }
+
+func isRetryableError(err error) bool {
+	if e, ok := err.(*googleapi.Error); ok && e.Code == 409 {
```
[the remark this answers is not in the document]
Comment (katrogan) [reply pkg/common/data_store.go:40]: I think this looks right, but it also looks like flytestdlib wraps the error: https://github.com/flyteorg/flytestdlib/blob/15cac7bec8470974050ed02343cda8191e5f14bc/storage/stow_store.go#L264 so you may need to unwrap the error. I think [IsCausedBy](https://github.com/flyteorg/flytestdlib/blob/15cac7bec8470974050ed02343cda8191e5f14bc/errors/error.go#L87) works here to get the underlying googleapi error.

Mind adding a unit test to verify this?

[the remark this answers is not in the document]
Comment (sonjaer) [reply pkg/common/data_store.go]: Fixed 👍

**Review by katrogan — APPROVED** [review]
